[X86] Support more ISAs to enable __builtin_cpu_supports

FreddyLeaf opened the pull request:
This patch will also expand supports for attribute/target, while
the priority of newly supported ISAs will be set to zero.

FreddyLeaf committed:
[X86] Support more ISAs to enable __builtin_cpu_supports

This patch will also expand supports for attribute/target, while
the priority of newly supported ISAs will be set to zero.

phoebewang commented on llvm/include/llvm/TargetParser/X86TargetParser.def:
```diff
@@ -173,85 +173,84 @@ X86_FEATURE_COMPAT(AVX512VNNI,      "avx512vnni",            34)
+X86_FEATURE_COMPAT(CMPXCHG16B,      "cx16",                   0)
+X86_FEATURE_COMPAT(CMPXCHG8B,       "cx8",                    0)
+X86_FEATURE_COMPAT(ENQCMD,          "enqcmd",                 0)
+X86_FEATURE_COMPAT(F16C,            "f16c",                   0)
```
Why clear the previous number?

FreddyLeaf replied:
first `f16c` should not be set higher priority than `avx512f`. higher priority means it can imply all the ISAs lower than it.

phoebewang commented on llvm/include/llvm/TargetParser/X86TargetParser.def:
```diff
@@ -173,85 +173,84 @@ X86_FEATURE_COMPAT(AVX512VNNI,      "avx512vnni",            34)
+X86_FEATURE       (FXSR,            "fxsr")
+X86_FEATURE_COMPAT(WIDEKL,          "widekl",                 0)
+X86_FEATURE_COMPAT(AVXVNNI,         "avxvnni",                0)
+X86_FEATURE_COMPAT(AVX512FP16,      "avx512fp16",             0)
```
ditto.

FreddyLeaf replied:
gcc only set priority before `avx512f`: https://github.com/gcc-mirror/gcc/blob/9693459e030977d6e906ea7eb587ed09ee4fddbd/gcc/common/config/i386/i386-cpuinfo.h#L149
which means, they don't support ISAs to function multiversion features(_-attribute__((target())) and so on) after `avx512f`, see this example: https://godbolt.org/z/Ec96ex3cK
The reason why they didn't extend the priority feature list can be found in this thread: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89929

> As discussed in https://gcc.gnu.org/ml/gcc-patches/2019-04/msg00416.html email thread, we reached the following consensus with H.J:
> 
> - As any AVX512 extensions (apart from AVX512F) can be enabled individually, it's difficult to come up with priorities in dispatcher.
> - We don't have a syntax for target_clone attribute where one would say e.g. avx512f+avx512cd+avx512er.
> - So that we would reject these (AVX512* except AVX512F) in target_clone attribute and we recommend to use rather.
> target_clones(arch=skylake,arch=skylake-avx512,arch=cannonlake,arch=icelake-client,arch=icelake-server, ..)
> - Using that one can cover used AVX512 ISA combinations for existing CPUs

FreddyLeaf committed:
clang-format and other refine

MaskRay commented on llvm/lib/TargetParser/X86TargetParser.cpp:
```diff
@@ -750,13 +750,16 @@ unsigned llvm::X86::getFeaturePriority(ProcessorFeatures Feat) {
-  std::iota(HelperList.begin(), HelperList.end(), 0);
+  const size_t MaxPriority = 37;
+  std::iota(HelperList.begin(), HelperList.begin() + MaxPriority + 1, 0);
+  for (int i = MaxPriority + 1; i != std::size(Priorities); ++i)
```
`int`=>`size_t`

FreddyLeaf replied:
Thanks review, will address in next update.

FreddyLeaf replied:
5922fc9

FreddyLeaf committed:
Fix lit fail of attr-cpuspecific.c

By removing the support of cx8 first, the current cpu_specific logic
requires the "generic" cpu to be with no CMPXCHG8B.

FreddyLeaf committed:
resovle compiler-rt build fail

FreddyLeaf commented:
ping for review

phoebewang commented on compiler-rt/lib/builtins/cpu_model/x86.c:
```diff
@@ -139,20 +139,77 @@ enum ProcessorFeatures {
   FEATURE_AVX512BF16,
   FEATURE_AVX512VP2INTERSECT,
+  FEATURE_3DNOW,
+  FEATURE_ADX = 40,
+  FEATURE_CLDEMOTE = 42,
```
Why do we skip some of features?

FreddyLeaf replied:
The missing features are ones I can't recognize in https://github.com/gcc-mirror/gcc/blob/02948ced062b730e2859f718dd46bfcaa3c16c6b/gcc/common/config/i386/i386-cpuinfo.h#L154, e.g. FEATURE_ABM, I can't find relate one in LLVM.

FreddyLeaf replied:
agree, done in 10c2a7c

phoebewang replied:
ABM is a AMD specific feature https://en.wikipedia.org/wiki/X86_Bit_manipulation_instruction_set
Maybe LLVM doesn't support it. But we should be able to map it to its corresponding [CPUID bit](https://github.com/gcc-mirror/gcc/blob/02948ced062b730e2859f718dd46bfcaa3c16c6b/gcc/config/i386/cpuid.h#L56). Maybe not in this patch.
I'd expect we can list the missing parts with comments like
```
FEATURE_3DNOW,
FEATURE_ADX = 40,
// FEATURE_ABM,
FEATURE_CLDEMOTE = 42,
```

FreddyLeaf replied:
Good idea. Addressed in 098a413

FreddyLeaf committed:
Address comments.

FreddyLeaf commented:
ping for review

MaskRay reviewed (approved):
LGTM, but this should be signed off by phoebewang

phoebewang commented on compiler-rt/lib/builtins/cpu_model/x86.c:
```diff
@@ -139,20 +139,79 @@ enum ProcessorFeatures {
+
+  FEATURE_ENQCMD = 48,
+  FEATURE_F16C,
+  FEATURE_FSGSBASE,
```
I didn't check others, but think `FEATURE_IBT` should map tp `FeatureSHSTK` in LLVM.

FreddyLeaf replied:
Thanks for checking. But gcc also has another `FEATURE_SHSTK`

phoebewang replied:
Oh, you are right. The 2 features map to two CPUID bits in SDM

> Bit 07: CET_SS. Supports CET shadow stack features if 1. Processors that set this bit define bits 1:0
> of the IA32_U_CET and IA32_S_CET MSRs. Enumerates support for the following MSRs: IA32_INTERRUPT_SPP_TABLE_ADDR, IA32_PL3_SSP, IA32_PL2_SSP, IA32_PL1_SSP, and IA32_PL0_SSP.

> Bit 20: CET_IBT. Supports CET indirect branch tracking features if 1. Processors that set this bit
> define bits 5:2 and bits 63:10 of the IA32_U_CET and IA32_S_CET MSRs.

FreddyLeaf committed:
Merge branch 'main' into more_isa

FreddyLeaf committed:
Address comments.

phoebewang commented on compiler-rt/lib/builtins/cpu_model/x86.c:
```diff
@@ -139,20 +139,88 @@ enum ProcessorFeatures {
+  FEATURE_SM3,
+  FEATURE_SHA512,
+  FEATURE_SM4,
+  // FEATURE_APX_F,
+  FEATURE_USERMSR = 112,
+  // FEATURE_AVX10_1_256,
+  // FEATURE_AVX10_1_512,
```
Why comments them? We have APX and AVX10 support.

FreddyLeaf replied:
addressed in 770ed9a

FreddyLeaf committed:
address comments.

phoebewang commented on compiler-rt/lib/builtins/cpu_model/x86.c:
```diff
@@ -217,10 +217,11 @@ enum ProcessorFeatures {
-  FEATURE_USERMSR = 112,
-  // FEATURE_AVX10_1_256,
-  // FEATURE_AVX10_1_512,
+  //FIXME: gcc used string "apxf" instead of "egpr"
```
The internal name is not important, so do not need a FIXME if we won't fix it.
I'm more concerning about the name in `(void)__builtin_cpu_supports("egpr");`
Does GCC use `egpr` too? Or we should change to "apxf"?

KanRobert replied:
GCC is using "apxf ". I believe we need to discuss with GCC team about this.

FreddyLeaf replied:
Yes, the concern is from that gcc use `apxf` but not `egpr`. I don't think adding one more line like this is complete safe. I prefer to not touch APX firs and refine it in another PR, WDYT?
```
X86_FEATURE_COMPAT(EGPR,            "apxf",                   0)
```

KanRobert commented on llvm/include/llvm/TargetParser/X86TargetParser.def:
```diff
@@ -173,85 +173,86 @@ X86_FEATURE_COMPAT(AVX512VNNI,      "avx512vnni",            34)
+X86_FEATURE_COMPAT(SM3,             "sm3",                    0)
+X86_FEATURE_COMPAT(SHA512,          "sha512",                 0)
+X86_FEATURE_COMPAT(SM4,             "sm4",                    0)
+X86_FEATURE_COMPAT(EGPR,            "egpr",                   0)
```
I suggest we can create another PR for APX features. What we should use for X86_FEATURE_COMPAT is not clear so far.

KanRobert commented on llvm/lib/TargetParser/Host.cpp:
```diff
@@ -1845,6 +1845,7 @@ bool sys::getHostCPUFeatures(StringMap<bool> &Features) {
   Features["prefetchi"]  = HasLeaf7Subleaf1 && ((EDX >> 14) & 1);
   Features["usermsr"]  = HasLeaf7Subleaf1 && ((EDX >> 15) & 1);
   Features["avx10.1-256"] = HasLeaf7Subleaf1 && ((EDX >> 19) & 1);
+  Features["egpr"] = HasLeaf7Subleaf1 && ((EDX >> 21) & 1);
```
Same here. Should discuss with GCC team and create a separate PR.

FreddyLeaf committed:
abandon apx related changes.

phoebewang commented on compiler-rt/lib/builtins/cpu_model/x86.c:
```diff
@@ -217,8 +217,7 @@ enum ProcessorFeatures {
-  //FIXME: gcc used string "apxf" instead of "egpr"
-  FEATURE_EGPR,
+  // FEATURE_APXF,
   FEATURE_USERMSR,
```
You need to change the number then.

FreddyLeaf committed:
address comment

phoebewang reviewed (approved):
LGTM.

FreddyLeaf commented:
Thanks review!
